src/algebra/module_regular.lean

```diff
@@ -0,0 +1,185 @@
+/-
+Copyright (c) 2021 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import algebra.group
+import algebra.group_power.basic
+import algebra.iterate_hom
+import algebra.module.basic
+
+/-!
+# Action of regular elements on a module
+
+We introduce `M`-regular elements, in the context of an `R`-module `M`.
+
+By definition, an `M`-regular element of a commutative ring acting on an `R`-module `M` is an
+element `a ∈ R` such that the function `m ↦ a • m` is injective.
+-/
+
+--should it be in the module namespace?
+
+namespace module
+
+variables {R S : Type*} {a b : R} {s : S} (M : Type*)
+
+section has_scalar
+
+variables [has_scalar R M] [has_scalar R S] [has_scalar S M] [is_scalar_tower R S M]
+
+/-- An `M`-regular element is an element `c` such that multiplication on the left by `c` is an
+injective map `M → M`. -/
+def is_regular (c : R) := function.injective ((•) c : M → M)
+
+section monoid
+
+--variables [monoid R]
+--semigroup does not have the right instance
+--instance semigroup.has_scalar : has_scalar R R := { smul := λ a b, a * b }
+
+/-- In a monoid, the product of `M`-regular elements is `M`-regular. -/
+lemma is_regular.mul (ra : is_regular M a) (rs : is_regular M s) :
+  is_regular M (a • s) :=
+λ a b ab, rs (ra ((smul_assoc _ _ _).symm.trans (ab.trans (smul_assoc _ _ _))))
+
+/--  If an element `b` becomes `M`-regular after multiplying it on the left by an `M`-regular
+element, then `b` is `M`-regular. -/
+lemma is_regular.of_mul (ab : is_regular M (a • s)) :
+  is_regular M s :=
+@function.injective.of_comp _ _ _ (λ m : M, a • m) _ (λ c d cd, ab
+  (by rwa [smul_assoc, smul_assoc]))
+
+/--  An element is `M`-regular if and only if multiplying it on the left by an `M`-regular element
+is `M`-regular. -/
+@[simp] lemma mul_is_regular_iff (b : S) (ha : is_regular M a) :
+  is_regular M (a • b) ↔ is_regular M b :=
+⟨λ ab, is_regular.of_mul M ab, λ ab, is_regular.mul M ha ab⟩
+
+variables [monoid R] [is_scalar_tower R R M]
```

Comment:
Don't you want to close the `has_scalar` section here?

Reply:
after removing the `section monoid` above:
```suggestion
end has_scalar

section monoid

variables [monoid R] [mul_action R M]
```

Reply:
Although actually, perhaps my claim is untrue...
```lean
/-- `mul_action` is equivalent to `is_scalar_tower` -/
example {R M : Type*} [monoid R] :
  mul_action R M ≃ (Σ' [has_scalar R M], by exactI is_scalar_tower R R M) :=
{ to_fun := λ h, by exactI ⟨infer_instance, infer_instance⟩,
  inv_fun := λ h,
  { mul_smul := λ r s m, let ⟨i1, i2⟩ := h in by resetI; rw [←smul_eq_mul, smul_assoc r s m],
    one_smul := λ m, let ⟨i1, i2⟩ := h in begin
      resetI,
      sorry, -- or is it
    end,
    ..h.1},
  left_inv := λ x, begin rcases x with ⟨⟨smul⟩⟩, refl, end,
  right_inv := λ x, begin rcases x with ⟨⟨smul⟩⟩, refl, end}
```

Reply:
Yes, leaving open the section `is_scalar` was a mistake!

As for your example, I do not really understand what is going on.  You are trying to prove that `mul_action R M` and `is_scalar_tower R R M` are compatible, I guess.  While I am lost on many of the steps, if just before the `sorry` you type `cases h`, you are left with 2 assumptions `has_scalar R M` and two `is_scalar_tower R R M`.  This seems a little unsettling.  Could this be the reason why you did not manage to finish the proof above?

Reply:
In fact, I think that in the `is_scalar_tower` definition, when you use `R R M`, you may use different `has_scalar R M` for the two slots to fill in and only require one to have `1` act as the identity, but not the other.  I am trying to see if I can get an example in which R = M = {0, 1} and the candidate actions are the "obvious one, where 0 absorbs and 1 is the identity and the other one where everything multiplies to 0.  However, I am not able to get Lean to process different instances on the same type...

Reply:
Here is an attempt: as you can see, there are plenty of combinations of good and bad `is_scalar_towers` on `{0, 1}` that work and only two that do not.  For some of these, at least one of the actions is good and it might be the one that is picked up by the `mul_action` thing.

```lean
import group_theory.group_action.defs

section counter

@[derive [decidable_eq]]
inductive z2 : Type
| zero : z2
| one : z2

namespace z2

instance : has_zero z2 := ⟨zero⟩
instance : has_one z2 := ⟨one⟩


/-- The "good action": zero acts as zero and one as the identity. -/
def ga : z2 → z2 → z2
| zero _ := zero
| _ zero := zero
| one one := one

/-- The "bad action": everything act as zero. -/
def ba : z2 → z2 → z2
| _ _ := zero

def mixbbb : @is_scalar_tower z2 z2 z2 { smul := ba } { smul := ba} { smul := ba} :=
{ smul_assoc := λ x y z, by cases x; cases y; cases z; dec_trivial }

def mixbbg : @is_scalar_tower z2 z2 z2 { smul := ba } { smul := ba} { smul := ga} :=
{ smul_assoc := λ x y z, by cases x; cases y; cases z; dec_trivial }

def mixbgb : @is_scalar_tower z2 z2 z2 { smul := ba } { smul := ga} { smul := ba} :=
{ smul_assoc := λ x y z, by cases x; cases y; cases z; dec_trivial }

-- does not work
def mixbgg : @is_scalar_tower z2 z2 z2 { smul := ba } { smul := ga} { smul := ga} :=
{ smul_assoc := λ x y z, by cases x; cases y; cases z; dec_trivial }

def mixgbb : @is_scalar_tower z2 z2 z2 { smul := ga } { smul := ba} { smul := ba} :=
{ smul_assoc := λ x y z, by cases x; cases y; cases z; dec_trivial }

def mixgbg : @is_scalar_tower z2 z2 z2 { smul := ga } { smul := ba} { smul := ga} :=
{ smul_assoc := λ x y z, by cases x; cases y; cases z; dec_trivial }

-- does not work
def mixggb : @is_scalar_tower z2 z2 z2 { smul := ga } { smul := ga} { smul := ba} :=
{ smul_assoc := λ x y z, by cases x; cases y; cases z; dec_trivial }

def mixggg : @is_scalar_tower z2 z2 z2 { smul := ga } { smul := ga} { smul := ga} :=
{ smul_assoc := λ x y z, by cases x; cases y; cases z; dec_trivial }

instance mz2 : monoid z2 :=
{ mul := ga,
  mul_assoc := λ a b c, by cases a;cases b;cases c; dec_trivial,
  one := 1,
  one_mul := λ a, by cases a; dec_trivial,
  mul_one := λ a, by cases a; dec_trivial }

def ma : mul_action z2 z2 :=
{ smul := (*),
  one_smul := λ a, by cases a; dec_trivial,
  mul_smul := λ a b c, by cases a; cases b; cases c; dec_trivial }
```

Reply:
> You are trying to prove that mul_action R M and is_scalar_tower R R M are compatible, I guess.

Not just compatible, but equivalent; since I made a claim to that effect in an earlier comment. I think you're right though, they're not equivalent.

> While I am lost on many of the steps, if just before the `sorry` you type `cases h`, you are left with 2 assumptions `has_scalar R M` and two `is_scalar_tower R R M`. This seems a little unsettling. Could this be the reason why you did not manage to finish the proof above?

Typing `cases h` is just repeating the work I did in the `let ⟨i1, i2⟩ := h` line, which is why it duplicates the typeclasses. This is a distraction.

> you may use different `has_scalar R M` for the two slots to fill in and only require one to have `1` act as the identity, but not the other.

This is an interesting question, but I think it's largely academic. In practice, mathlib uses type aliases if it wants to attach a different action to a type, so this wouldn't actually come up.